runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,304 @@
+     */
+    public static class CorfuQueueRecord<E> implements Comparable<CorfuQueueRecord<? extends E>> {
+        /**
+         * This ID represents the entry and its order in the Queue.
+         * This implies that it is unique and comparable with other IDs
+         * returned from CorfuQueue methods with respect to its enqueue order.
+         * However it cannot be compared with ID returned from the enqueue method for ordering
+         * because if this method is wrapped in a transaction, the order is established only later.
+         */
+        @Getter
+        private final CorfuRecordId recordId;
+
+        @Getter
+        private final E entry;
+
+        public String toString() {
+            return String.format("%s=>%s", recordId, entry);
+        }
+
+        CorfuQueueRecord(long ordering, long entryId, E entry) {
+            this.recordId = new CorfuRecordId(ordering, entryId);
+            this.entry = entry;
+        }
+
+        @Override
+        public int compareTo(CorfuQueueRecord<? extends E> o) {
+            return this.recordId.compareTo(o.getRecordId());
+        }
+    }
+
+    /**
+     * We need to encode 2 pieces of information into an 8 byte long to represent ordering.
+     * 1. We use 40 bits for CorfuQueue's snapshot version information.
+     * 2. We use 24 bits for the index within that CorfuQueue's snapshot.
+     */
+    final private static int MAX_BITS_FOR_INDEX = 24;
+
+    /**
+     * Returns a List of CorfuQueueRecords sorted by the order in which the enqueue materialized.
+     * This is the primary method of consumption of entries enqueued into CorfuQueue.
+     *
+     * To re-constitute the commit order in the presence of transactions the CorfuRecords returned
+     * here have their UUID fields composed of two parts:
+     *     +----------------------------------------------------------------------------+
+     *     | Stream snapshot |Index in snapshot|     ID of the entry in the map        |
+     *     +----------------------------------------------------------------------------+
+     *     <----5 bytes------><----3 bytes-----><------- 8 bytes ----------------------->
+     *
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntries - Limit the number of entries returned from start of the queue
+     * @throws IllegalArgumentException if maxEntries is negative.
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
+        if (maxEntries < 0) {
+            throw new IllegalArgumentException("entryList given negative maxEntries");
+        }
+
+        // Bind the iteration order to a snapshot of the Queue using a transaction.
+        long snapshotVersion;
+        boolean startedNewTransaction = false;
+        if (TransactionalContext.isInTransaction()) {
```

Comment:
Yes, thanks for the observation @WenbinZhu , please note that this is solved with the CorfuRecordId class's equals() function which only matches the LSB (the Id part), so in your example id1 from t1 and id2 from t2 will match to be equal. Also this is something we discussed quite a bit. The original plan was to return the exact stream addresses, but that would break layer abstraction, so it was decided to return ids in this manner.

Reply:
Yes, if we use euqals() I think it's fine, but if we care about  ordering and use compareTo(), <99, 1, id2> will be smaller than <100, 0, id1>. Does that break your semantics?